modules/parent-join/src/main/java/org/elasticsearch/join/aggregations/ChildrenToParentAggregator.java

```diff
@@ -0,0 +1,102 @@
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.bucket.BucketsAggregator;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link BucketsAggregator} which resolves to the matching parent documents.
+ *
+ * It ensures that each parent only matches once per bucket.
+ */
+public class ChildrenToParentAggregator extends AbstractParentChildAggregator {
+
+    public ChildrenToParentAggregator(String name, AggregatorFactories factories,
+            SearchContext context, Aggregator parent, Query childFilter,
+            Query parentFilter, ValuesSource.Bytes.WithOrdinals valuesSource,
+            long maxOrd, List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData)
+            throws IOException {
+        super(name, factories, context, parent, childFilter, parentFilter,
+            valuesSource, maxOrd, pipelineAggregators, metaData);
+    }
+
+    @Override
+    public InternalAggregation buildAggregation(long owningBucketOrdinal) throws IOException {
+        return new InternalParent(name, bucketDocCount(owningBucketOrdinal),
+                bucketAggregations(owningBucketOrdinal), pipelineAggregators(), metaData());
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        return new InternalParent(name, 0, buildEmptySubAggregations(), pipelineAggregators(),
+                metaData());
+    }
+
+    @Override
+    LeafBucketCollector getLeafBucketCollector(SortedSetDocValues globalOrdinals, Bits childDocs) {
+        return new LeafBucketCollector() {
+
+            @Override
+            public void collect(int docId, long bucket) throws IOException {
+                if (childDocs.get(docId) && globalOrdinals.advanceExact(docId)) {
+                    long globalOrdinal = globalOrdinals.nextOrd();
+                    assert globalOrdinals.nextOrd() == SortedSetDocValues.NO_MORE_ORDS;
+                    if (globalOrdinal != -1) {
+                        long bucketPrev = ordinalToBuckets.get(globalOrdinal);
+
+                        // add bucket if none is set
+                        if (bucketPrev == -1) {
+                            ordinalToBuckets.set(globalOrdinal, bucket);
+                        } else if (bucketPrev != bucket) {
+                            // otherwise store it in the array, but only if the bucket is different than
+                            // the previous one for this childDoc, as we only need one occurrence
+                            // of child -> parent for this aggregation
```

Comment:
As explained above this part is problematic. It is required to deduplicate the bucket per global ordinal but the cost to do this seems prohibitive. I discussed with @martijnvg offline and he proposed to restrict this new aggregation to the root level (would throw an exception if a parent aggregation is defined). I like this idea because we can assume the same cost for the `parent => children` and `children => parent` cases, we just need an array of global ordinals and `bucket` can be ignored completely.

Reply:
You mean the previous version (https://github.com/elastic/elasticsearch/pull/34210/commits/b6e186cdea45e0987082ba4328e38f93898c8f39#diff-e268a24051f30621c407964faaa1fb24) was correct and this one is not? I verified that all cases that I could think of were handled correctly. Can you sketch out what constellation of aggregations/data would not work here so I can create a test case which verifies it?

My use-case calls for a general implementation as we would like to provide a very flexible search functionality on top of it, but having it enabled for certain uses would be a start. We also experimented with adding a custom plugin/module with the aggregation, so we have a way to go forward with something while having Elasticsearch 6.x in use, so this PR is more a long-term way to integrate this into Elasticsearch itself.

On performance, some rough tests did not show a huge performance problem with this aggregation, but naturally your view of requirements here is much broader, which cases did you think of that will cause issues?